src/Components/test/Microsoft.AspNetCore.Components.Test/Rendering/HtmlRendererTests.cs

```diff
@@ -0,0 +1,425 @@
+
+
+        [Fact]
+        public void RenderComponent_DoesNotEncodeMarkup()
+        {
+            // Arrange
+            var expectedHtml = new[] { "<", "p", ">", "<span>Hello world!</span>", "</", "p", ">" };
+            var serviceProvider = new ServiceCollection().AddSingleton(new RenderFragment(rtb =>
+            {
+                rtb.OpenElement(0, "p");
+                rtb.AddMarkupContent(1, "<span>Hello world!</span>");
+                rtb.CloseElement();
+            })).BuildServiceProvider();
+            var htmlRenderer = new HtmlRenderer(serviceProvider, HtmlEncoder.Default);
+
+            // Act
+            var result = htmlRenderer.RenderComponent<TestComponent>(ParameterCollection.Empty);
+
+            // Assert
+            Assert.Equal(expectedHtml, result);
+        }
+
+
+        [Fact]
+        public void RenderComponent_CanRenderWithAttributes()
+        {
+            // Arrange
+            var expectedHtml = new[] { "<", "p", " ", "class", "=", "\"", "lead", "\"", ">", "Hello world!", "</", "p", ">" };
+            var serviceProvider = new ServiceCollection().AddSingleton(new RenderFragment(rtb =>
+            {
+                rtb.OpenElement(0, "p");
+                rtb.AddAttribute(1, "class", "lead");
+                rtb.AddContent(2, "Hello world!");
+                rtb.CloseElement();
+            })).BuildServiceProvider();
+
+            var htmlRenderer = new HtmlRenderer(serviceProvider, HtmlEncoder.Default);
+
+            // Act
+            var result = htmlRenderer.RenderComponent<TestComponent>(ParameterCollection.Empty);
+
+            // Assert
+            Assert.Equal(expectedHtml, result);
+        }
+
+        [Fact]
+        public void RenderComponent_HtmlEncodesAttributeValues()
+        {
+            // Arrange
+            var expectedHtml = new[] { "<", "p", " ", "class", "=", "\"", "&lt;lead", "\"", ">", "Hello world!", "</", "p", ">" };
+            var serviceProvider = new ServiceCollection().AddSingleton(new RenderFragment(rtb =>
+            {
+                rtb.OpenElement(0, "p");
+                rtb.AddAttribute(1, "class", "<lead");
+                rtb.AddContent(2, "Hello world!");
+                rtb.CloseElement();
+            })).BuildServiceProvider();
+
+            var htmlRenderer = new HtmlRenderer(serviceProvider, HtmlEncoder.Default);
+
+            // Act
+            var result = htmlRenderer.RenderComponent<TestComponent>(ParameterCollection.Empty);
+
+            // Assert
+            Assert.Equal(expectedHtml, result);
+        }
+
+        [Fact]
+        public void RenderComponent_CanRenderBooleanAttributes()
+        {
+            // Arrange
+            var expectedHtml = new[] { "<", "input", " ", "disabled", " />" };
+            var serviceProvider = new ServiceCollection().AddSingleton(new RenderFragment(rtb =>
+            {
+                rtb.OpenElement(0, "input");
+                rtb.AddAttribute(1, "disabled", true);
```

Comment:
Could this test also add a separate `false` attribute to show it doesn't appear in the resulting HTML?